### PR TITLE
Keep Subsequent identifier links on the same line.

### LINKED ISSUE #2712: Subsequent identifier links in chain should be on the same line
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/main?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAIgK7QDWAFAJRgC8AOgE60B2EEAtomwEaL1joALAJYBnAHRFSEagWIkwlBs1ZgO3Xv2HiA4rFgjElGrNKLGLVkqYAbROjDxjk8hStX4EuZ9LeSY3fqGFL7%2BegaUIc6hgRFRcV4B4cHxPil%2BiUFu5gD02WAAogAeuNDoiPDukV7OVirRSbWsIRlmyk1p9ZnmdR1RLa7d7dXDqQlhQZAANOBQBujoQkwA5iKoYADaMxCgKhCQACRYuGuQi%2BhTW6wHAG4AhtaEyGibu6wADNOvewCSTEhM6AAykIAF7IT5fAAslwgAF1LgBfCHbGFQQ44J5QM4XV43e6PNYvL4ARmRu0gAFlboUADKLRA0xDLQQ4kkAJjeqPhKiRlx25PRxzQkCZ8AA8gAzOlMRCArC2VkqPEPTFE15sslK-J-SXS8GovbWCWQLmIsn8pWCzGQLh6ayKq77O4qwkGsAAZk1V1%2BBno6AAYot7gA5RAAd2s9IdKnQ9Eepp55tRByO1ttsHtXuVBOebshXr2gOwt2giAAQogJbB6IgAAq3ei3DhlejR1ix%2BOvbmsXkqC2O1MndOZ5NO-Gqt0AVgLUCLJfLlerDNgYd40FuBl%2B11g64WsCYbYgHcQCZ7SdxVqHdrb2YnXzAADYZ5A56WK1WawBVbC4ejrzdMNuu5CPuh5gBK9wGKeEC9qw-Z7JewrDjeY4urm94AOzPq%2BC4fogADC1gbiI%2BH7iIHalNWYEQdYUFdmafKjoOSHXlmqE5hsboABzYcWb6LjWFKcDwrYzjRdG7N2MHngKzFQMhbHOhxaq7AAnLx87vkupHWKBYmQSe9GJoxF5yTarGjkpd4kh8bovnxiAAIISi2pFsGwtxgce0FgLBKKmRiV4ZihVmuvexKknZOFaTWgKcEIcC6Qe%2Bm0YZkkMX2TGBSxwWKeOYXshppbOS2cVsAlGZ6W63lGWeJmydl8kWReoXoSSnpRQ5jn0LAhB-PgiCRuVLZeXGaUqFJvkyZaZnYnlaGceF%2BZ2VShTfBKAAqAhMoCAjVugADqQjwCyM6crV0n1TNjWnACIX5W1rzEtOK3UutW1MvkqW7ftR0nQIYEPud6XGZlAVClid3zcpbrEk%2Br1rUwEpCIUYq-rc6DVkU2A1iIIggcl3HAxNGVwVlEO3ec0PWU9WEIwASogcD0PAf2nXmxOsJNflgPBaKzVDlkPYtJI8QzTPVvAwaEOo9CSt8ZRsKsM7Ej5PN8ymN1sIQ1gLJGMr%2BtWHnzLwG2BdTBVPepdmM8z8AUjrev0ob9DGyNz7QAIDYli2AD6YbHSyavTQON1zULC0qSoHLPqt3WNpgYr0HS5FswDM5cZzcKk-5DUU%2BHLXC1HrBspF96UtS8e3InyeiOg0uy-LivK7DwdXaHFPa7rQj64gLtu6b5sRzD95shqdlVzXKfoA73e9-3GPu3Znve6UvD%2B4HANt2DefWgXAqtSL6odeXq0AGrjmWizwIsSxp2Bmfb2T4N74LheR26bLLaf1L%2Bn1pQEyvn8W%2B98ZyQizmAbmIcEICypsPGmuw2QvR-oUfAsB0DaDsNjXG%2BN9ygKJk-XO10KYKXgZbRB8Ny6hgjPSCs6BVxMjNrgAayMmBCD3EwRyfwhKyxbveGqIM6o72IWmZqB8i6fzpuXRykYlhMD-kwAB%2B5gRyIxoQGsG1YC-H%2BOgDGBNqIGUIbzcmojcpkMeogsW0jda8CYHo64y5lg8JEiwxY7CCZ8K%2BOJcaXMc7GJfkFEc78R5fDZNbU%2BjshB6hpE2Lg8BbiEX0LfahvcDGpSMRrRCTUzHBIQSod0tly7Y14EIDgAJ7gAGlECIGwNopk6BfhlkbIo9ObpvEZJMYE%2B6H97zujLl8SAZYiJMBIHqEQ3Vep-FDORcos8nYymwYgPGHjRqdkEZdYRHdTFBPET0r47px7lzLA2GKiB8CiGgPQUpQYyjwE-Gw-cA1oBEUbBwtJEkSag2frvE4Xd5nlkbKQOwcpMAKgthY-JJ8BlzJ7rQwFJBgXyn1OXS5TZcDwBNBdKa7cYFhzfrskJrx3TfwGVUmpq0G4iUlEM24IyxlgWJG8CBUCcX8xuqQ3J5D8nIIGXKK5pQKQIGRV4wxWKEQzFhBCU4Ih-T4zWO0hEQA)

#### Code

```fsharp
type Duck() =
    member this.Duck  = Duck ()
    member this.Goose() = Duck()
    
let d = Duck()

d.Duck.Duck.Duck.Goose().Duck.Goose().Duck.Duck.Goose().Duck.Duck.Duck.Goose().Duck.Duck.Duck.Duck.Goose()
```
            

#### Result

```fsharp
type Duck() =
    member this.Duck = Duck()
    member this.Goose() = Duck()

let d = Duck()

d.Duck.Duck.Duck
    .Goose()
    .Duck.Goose()
    .Duck.Duck
    .Goose()
    .Duck.Duck
    .Duck
    .Goose()
    .Duck.Duck
    .Duck
    .Duck
    .Goose()
```
            
#### Problem description

I was expecting:

```fsharp
d.Duck.Duck.Duck
    .Goose()
    .Duck.Goose()
    .Duck.Duck.Goose()
    .Duck.Duck.Duck.Goose()
    .Duck.Duck.Duck.Duck.Goose()
```

I believe the chain is correctly constructed but not formatted rightly in CodePrinter.

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas main branch at 2023-01-12T20:47:35Z - b6f7ecfd364a4db5a67888acc3248c37f1424c37

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>

### PR DESCRIPTION
**Pull request by nojaf**
Fixes #2712.

### CODE REVIEW

File: src/Fantomas.Core.Tests/ChainTests.fs [hunk 1]
```diff
@@ -299,3 +299,52 @@ Map
...
+d.Duck.Duck.Duck
+    .Goose()
+    .Duck.Goose()
+    .Duck.Duck.Goose()
+    .Duck.Duck.Duck.Goose()
+    .Duck.Duck.Duck.Duck.Goose()
+"""
+
+[<Test>]
+let ``very long chain with a some index expressions`` () =
```
Comment (nojaf) [inline src/Fantomas.Core.Tests/ChainTests.fs:336]: @dawedawe this one is a bit tricky. The whole chain does not fit within the `MaxDotGetExpressionWidth` boundary, but there is no application to chop it logically.

Comment (nojaf) [reply src/Fantomas.Core.Tests/ChainTests.fs:336]: It follows the code path where the leading links are all the links in
```fsharp
match links with
| [] ->
    fun ctx ->
        isShortExpression
            ctx.Config.MaxDotGetExpressionWidth
            short
            (match leadingChain with
             | [] -> sepNone
             | head :: links -> genFirstLinkAndIndentOther head links)
            ctx
```

The `head :: links -> genFirstLinkAndIndentOther head links` will print the first on a single line as a last resort.

This feels a bit distorted. Thoughts?

File: src/Fantomas.Core/CodePrinter.fs [hunk 2]
```diff
@@ -1123,7 +1129,7 @@ let genExpr (e: Expr) =
                             short
                             (match leadingChain with
                              | [] -> sepNone
-                             | head :: links -> genFirstLinkAndIndentOther head links)
+                             | head :: links -> genLink false head +> indent +> genIndentedLinks true links +> unindent)
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:1132]: @dawedawe I changed the implementation to only consider the `max_line_length` when there are only simple links in the chain.

You basically print the first one, if the second fits after it, great, if not put it on the next line. I think this is reasonable.

Comment (dawedawe) [reply src/Fantomas.Core/CodePrinter.fs:1132]: Yep, very reasonable.

File: src/Fantomas.Core.Tests/ChainTests.fs [hunk 3]
```diff
@@ -299,3 +299,75 @@ Map
...
+    |> prepend newline
+    |> should
+        equal
+        """
+Universe.Galaxy.SolarSystem.Planet.[3].Countries
+    .[9].People.Count
+"""
+
+[<Test>]
+let ``Even longer chain with only simple links`` () =
```
Comment (dawedawe) [inline src/Fantomas.Core.Tests/ChainTests.fs:354]:
```suggestion
let ``even longer chain with only simple links`` () =
```

**Review by dawedawe — APPROVED** [review]

**Pull request merged**